File: Week 6 - Python/sentimental-credit/credit.py
from cs50 import get_int
from cs50 import get_string


def main():
    card_nb = card_number()
    #print(f"CB : {card_nb}")

    count_digit = number_of_digit(card_nb)
    #print(f"count : {count_digit}")

    sum_total = checksum(count_digit, card_nb)
    #print(f"sum : {sum_total}")

    validation(sum_total, count_digit, card_nb)


def card_number():
    """ Prompt user for card number """
    card_nb = get_int("Credit Card Number : ")
    return card_nb


def number_of_digit(card_nb):
    """ count how many digit there is in user card number """
    # Initialize count_digit
    count_digit = 0

    while True:
        # running loop until card_nb = 0
        if card_nb == 0:
            break
        else:
            # // ==> to get an int (getting rid of numbers after comma) and not a float
            card_nb = card_nb // 10
            count_digit = count_digit + 1
    # Return how many digit there is in card number provide by user
    return count_digit


def checksum(count_digit, card_nb):

    sum10 = 0
    digit10 = 0
    digitX2 = 0
    digit_sum = 0
    digit = 0

    """ sum digit from card number """
    for i in range(1, count_digit+1):
        # EVEN
        # multiply each digits by 2 (starting with the number’s second-to-last digit)
        if i % 2 == 0:
            # %10 return last digit of card_nb
            digit = card_nb % 10

            # example : if digit * 2 = 12 we have to sum 1 + 2
            if digit * 2 >= 10:
                while True:
                    # running loop until digit = 0
                    if digit == 0:
                        break
                    else:
                        digit10 = digit * 2 % 10
                        digit = digit // 10
                        # + 1 because digit * 9 can't be greater than 18 so 18 %10 = 8
                        sum10 = digit10 + 1

                digitX2 = digitX2 + sum10

            else:
                digitX2 = digitX2 + 2 * digit
        # ODD
        # sum of the digits that weren’t multiplied by 2 (starting from the end)
        else:
            digit = card_nb % 10
            digit_sum = digit_sum + digit

        card_nb = card_nb // 10

    sum_total = digitX2 + digit_sum
    return sum_total


def validation(sum_total, count_digit, card_nb):
    """ check for valid card number (Mastercard/VISA/American Express/INVALID) """
    # Checking if card number is valid
    if sum_total % 10 == 0:

        digit = 0
        digit1 = 0

        if count_digit == 13 or count_digit == 15 or count_digit == 16:
            # CHECKING FOR AMERICAN EXPRESS
            if count_digit == 15:
                while True:
                    if card_nb == 0:
                        break
                    else:
                        digit = card_nb % 10

                        if card_nb <= 99 and card_nb >= 10:
                            digit1 = digit1 + digit
                        if card_nb <= 9:
                            digit1 = digit1 + digit * 10

                        card_nb = card_nb // 10

                if digit1 == 34 or digit1 == 37:
                    print("AMEX")
                else:
                    print("INVALID")

            # CHECKING FOR MASTERCARD
            if count_digit == 16:

                while True:
                    if card_nb == 0:
                        break
                    else:
                        digit = card_nb % 10

                        if card_nb <= 99 and card_nb >= 10:
                            digit1 = digit1 + digit
                        if card_nb <= 9:
                            digit1 = digit1 + digit * 10
                        card_nb = card_nb // 10

                #print(f"digit1: {digit1}")
                if digit1 == 51 or digit1 == 52 or digit1 == 53 or digit1 == 54 or digit1 == 55 or digit == 4:

                    if digit1 == 51 or digit1 == 52 or digit1 == 53 or digit1 == 54 or digit1 == 55:
                        print("MASTERCARD")

                    if digit == 4:
                        print("VISA")
                else:
                    print("INVALID")

            # CHECKING FOR VISA
            if count_digit == 13:

                while True:
                    if card_nb == 0:
                        break
                    else:
                        digit = card_nb % 10
                        card_nb = card_nb // 10

                if digit == 4:
                    print("VISA")
                else:
                    printf("INVALID")
        else:
            print("INVALID")

    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: Week 1 - C /credit/credit.c
#include <cs50.h>
#include <stdio.h>

long card_number(void);
int number_of_digit(long card_nb);
int checksum(int count_digit, long card_nb);
void validation(int sum_total, int count_digit, long card_nb);

int main(void)
{
    // prompt user for card number
    long card_nb = card_number();

    // Calculate the number of digit
    int count_digit = number_of_digit(card_nb);
    //printf("count_digit: %i\n", count_digit);

    // LOOP to verify validity of card number
    int sum_total = checksum(count_digit, card_nb);
    //printf("sum_total: %i\n", sum_total);

    // Validation        // identification du type de carte (AMEX, VISA ou MASTERCARD)
    validation(sum_total, count_digit, card_nb);
}

long card_number(void)
{
    long card_nb = get_long("Credit Card Number : "); // Ask for CB number
    return card_nb;
}

int number_of_digit(long card_nb)
{
    int count_digit = 0;
    do
    {
        // %10 return last digit from CB number input
        int digit = card_nb % 10;
        card_nb = card_nb / 10;
        count_digit = count_digit + 1;
    }
    while (card_nb > 0);
    return count_digit; // return how many digit there is in number provide by user
}

int checksum(int count_digit, long card_nb)
{
    int sum10 = 0;
    int digit10 = 0;
    int digitX2 = 0;
    int digit_sum = 0;
    int digit = 0;

    for (int i = 1; i <= count_digit; i++)
    {
        if (i % 2 == 0) // EVEN (from one before last digit) // sum of digit * 2
        {
            digit = card_nb % 10;

            // if digit*2 >= 10 // example : digit*2 = 12 we have to to 1 + 2
            if (digit * 2 >= 10)
            {
                do
                {
                    digit10 = digit * 2 % 10;
                    digit = digit / 10;
                    sum10 = digit10 + 1;
                }
                while (digit > 0);
                digitX2 = digitX2 + sum10;
            }
            else
            {
                digitX2 = digitX2 + 2 * digit;
            }
        }
        else // ODD (from last digit) sum of digit
        {
            digit = card_nb % 10;
            digit_sum = digit_sum + digit;
        }
        card_nb = card_nb / 10;
    }
    int sum_total = digitX2 + digit_sum;
    return sum_total;
}

void validation(int sum_total, int count_digit, long card_nb)
{
    if (sum_total % 10 == 0)
    {
        int digit = 0;
        int digit1 = 0;

        if (count_digit == 13 || count_digit == 15 || count_digit == 16)
        {
            // checking for American Express
            if (count_digit == 15)
            {
                do
                {
                    digit = card_nb % 10;

                    if (card_nb <= 99 && card_nb >= 10)
                    {
                        digit1 = digit1 + digit;
                    }
                    if (card_nb <= 9)
                    {
                        digit1 = digit1 + digit * 10;
                    }
                    card_nb = card_nb / 10;
                }
                while (card_nb > 0);

                if (digit1 == 34 || digit1 == 37)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }

            // checking for MasterCard
            if (count_digit == 16)
            {
                do
                {
                    digit = card_nb % 10;

                    if (card_nb <= 99 && card_nb >= 10)
                    {
                        digit1 = digit1 + digit;
                    }
                    if (card_nb <= 9)
                    {
                        digit1 = digit1 + digit * 10;
                    }
                    card_nb = card_nb / 10;
                }
                while (card_nb > 0);

                if (digit1 == 51 || digit1 == 52 || digit1 == 53 || digit1 == 54 || digit1 == 55 || digit == 4)
                {
                    if (digit1 == 51 || digit1 == 52 || digit1 == 53 || digit1 == 54 || digit1 == 55)
                    {
                        printf("MASTERCARD\n");
                    }
                    if (digit == 4)
                    {
                        printf("VISA\n");
                    }
                }
                else
                {
                    printf("INVALID\n");
                }
            }

            // checking for VISA
            if (count_digit == 13)
            {
                do
                {
                    digit = card_nb % 10;
                    card_nb = card_nb / 10;
                }
                while (card_nb > 0);

                if (digit == 4)
                {
                    printf("VISA\n");
                }
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
